### ISSUE TITLE
cmake gets confused on systems that have both qt4 and qt5 installed (Trac #477)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#476]
(Assigned this to nega, but not sure if he is the right guy to bother with this. We need someone who knows both about cmake and about Qt. And sjackso is leaving IceCube, unfortunately.)

My laptop has both Qt4 and Qt5 installed. When I run cmake then I see that Qt5 is found instead of Qt4. Since the include files are organized in a different way for Qt5, the build (steamshovel and glshovel, or anything else that depends on qt4) fails.

I made it work with a hack, namely by inserting the directory path where (on my system) the qt4 binaries are installed:

/usr/lib/x86_64-linux-gnu/qt4/bin

I inserted this path in cmake/tools/qt4.cmake just after the line with the keyword "PATHS" and before the line with "${I3_PORTS}/qt-4.6.0/bin".

That causes the build to complete, so now I can enjoy the functionality steamshovel again, but as already mentioned, this solution is a hack, I'm sure that there will be more users with both Qt5 and Qt4 on their systems, and it would be nice to fix it for real.

It looks like the Qt team has already foreseen this situation, because the "/usr/bin/qmake" program on my system is a symlink to /usr/bin/qtchooser, which can be called with an option -qt=qt4 or -qt=qt5 to enforce a particular Qt version. But I can't figure out where in cmake/tools/qt4.cmake (or in the CMakeLists.txt files) I could insert this. I tried adding this option setting to QT_DEFINITIONS, but that did not do anything. I failed to find out where qmake is actually called. I also checked these files:

```text
$I3_PORTS/share/cmake-2.8.5/Modules/UseQt4.cmake
/usr/share/cmake/Modules/UseQt4.cmake
```

And in neither of them I could find something that I recognized as something that would let me enforce qt4 in a meaningful way.

There probably more routes to try, but I'm running out of time, hence I'm dumping my findings so far in this ticket.

Migrated from https://code.icecube.wisc.edu/ticket/477
```json
{
    "status": "closed", 
    "changetime": "2019-01-11T23:10:47", 
    "description": "(Assigned this to nega, but not sure if he is the right guy to bother with this. We need someone who knows both about cmake and about Qt. And sjackso is leaving IceCube, unfortunately.)\n\nMy laptop has both Qt4 and Qt5 installed. When I run cmake then I see that Qt5 is found instead of Qt4. Since the include files are organized in a different way for Qt5, the build (steamshovel and glshovel, or anything else that depends on qt4) fails.\n\nI made it work with a hack, namely by inserting the directory path where (on my system) the qt4 binaries are installed:\n\n/usr/lib/x86_64-linux-gnu/qt4/bin\n\nI inserted this path in cmake/tools/qt4.cmake just after the line with the keyword \"PATHS\" and before the line with \"${I3_PORTS}/qt-4.6.0/bin\".\n\nThat causes the build to complete, so now I can enjoy the functionality steamshovel again, but as already mentioned, this solution is a hack, I'm sure that there will be more users with both Qt5 and Qt4 on their systems, and it would be nice to fix it for real.\n\nIt looks like the Qt team has already foreseen this situation, because the \"/usr/bin/qmake\" program on my system is a symlink to /usr/bin/qtchooser, which can be called with an option -qt=qt4 or -qt=qt5 to enforce a particular Qt version. But I can't figure out where in cmake/tools/qt4.cmake (or in the CMakeLists.txt files) I could insert this. I tried adding this option setting to QT_DEFINITIONS, but that did not do anything. I failed to find out where qmake is actually called. I also checked these files:\n\n{{{\n$I3_PORTS/share/cmake-2.8.5/Modules/UseQt4.cmake\n/usr/share/cmake/Modules/UseQt4.cmake\n}}}\n\nAnd in neither of them I could find something that I recognized as something that would let me enforce qt4 in a meaningful way.\n\nThere probably more routes to try, but I'm running out of time, hence I'm dumping my findings so far in this ticket.", 
    "reporter": "boersma", 
    "cc": "steve.jackson@icecube.wisc.edu, david.schultz@icecube.wisc.edu", 
    "resolution": "fixed", 
    "_ts": "1547248247244836", 
    "component": "steamshovel", 
    "summary": "cmake gets confused on systems that have both qt4 and qt5 installed", 
    "priority": "normal", 
    "keywords": "qt4 cmake steamshovel glshovel", 
    "time": "2013-12-18T11:42:15", 
    "milestone": "", 
    "owner": "nega", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20131219T18:07:45: boersma commented:

> It looks like one way to fix it is to use the "QT_SELECT" environment variable (giving it the value "qt4"). I need to do some more testing, but it looks like this solves it. The nice thing about this solution is that on mixed qt4/qt5 systems this would fix it without having to hack the cmake scripts.

**Comment by icecube-trac** [comment]
Trac update at 20141016T18:46:55: 
* richman changed attachment from "" to "davino_qt4.cmake_patch.patch"
* richman commented:

> patch from boersma that Works For Me

**Comment by icecube-trac** [comment]
Trac update at 20141016T18:47:47: david.schultz changed cc from "steve.jackson@icecube.wisc.edu" to "steve.jackson@icecube.wisc.edu, david.schultz@icecube.wisc.edu"

**Comment by icecube-trac** [comment]
Trac update at 20141204T20:34:18: 
* nega changed status from "new" to "closed"
* nega commented:

> In changeset 2013:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2013"
> stop getting confused between qt4 and qt5
> 
> fixes #476
> ```


* nega changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T23:10:47: nwhitehorn commented:

> In changeset 169907:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="169907"
> stop getting confused between qt4 and qt5
> 
> fixes #476
> ```
